### ISSUE TITLE
Beginner's Question: How to send traffic to pod in default configuration?

### ISSUE THREAD

**Issue by mkarg** [kubernetes-sigs/kind#3348]
First of all, thanks a lot to this community to bring us kind. Kind is awesome! 🚀 

I am used to Docker since several years, and now like to use kind to learn k8s. Sorry if my question is dumb, but I cannot find the answer in the k8s and kind docs.

I followed the instructions and set up kind on docker on Linux server host (no Docker Desktop, no GUI, just bash and docker) successfully. Then I did `kind create cluster`. The cluster is visible as a single Docker container exposing a single ephemerial port at 127.0.0.1 (`kind get clusters`, `docker ps`). Using cURL on that host I was able to successfully ask the cluster for its status using that address. So far, so good. 😃 

Next I did `kubectl apply -f echo.yml` to deploy the echo server using the yml example from https://kind.sigs.k8s.io/docs/user/configuration/. Using `kubectl get pods -o=wide` I can see the IP of the created pod. Great! 😃 

But when I execute `cURL http://<pod-ip>` I do not get any answer; instead the server could not get connected. The reason seems to be obvious: The sole existing docker container (the control plane) does not expose port 80, so no host, not even localhost, can connect into that docker container. 😢 

Maybe the solution seems to be obvious and I just miss to find it in the docs, but could anybody be so kind to simply tell me what exactly I need to do to allow cURL on the docker host to access the echo server running on the sole pod? 🤔 

That would be great! Thanks a lot in advance! kind is really cool! 👍

**Comment by aojea** [comment]
This is a routing problem, the pod has an IP that is internal to the container, so you need to use any technique that allows you to communicate with that internal IP, you can use port mapping https://kind.sigs.k8s.io/docs/user/configuration/#extra-port-mappings or ingress https://kind.sigs.k8s.io/docs/user/ingress , per example

**Comment by mkarg** [comment]
@ajayk Thank you for your kind help. Unfortunately it does not work still. I have recreated the cluster using `extraPortMappings`, so `docker ls` now tells me that the host port 888 on 127.0.0.1 is mapped (I checked that using `kubectl describe pod <podname>` and `docker ps`). So it *should* work now, but actually cURL still fails to connect:
```
*   Trying 127.0.0.1:888...
* Connected to 127.0.0.1 (127.0.0.1) port 888 (#0)
> GET / HTTP/1.1
> Host: 127.0.0.1:888
> User-Agent: curl/7.88.1
> Accept: */*
>
* Recv failure: Connection reset by peer
* Closing connection 0
curl: (56) Recv failure: Connection reset by peer
```
I though maybe port 888 is a problem, so I deleted the pod and tried the very same using docker directly, using `docker run -p 888:<containerport> <image>` and cURL was able to successfully access the echo server (this proofs that neither docker nor the image nor cURL nor the port mapping is the problem, but the cause must be somewhere between the pod and the mapped port).

Could it be the case that I have to tell kubernetes explicitly that the pod shall expose its port (I assume kubernetes does not consider the EXPOSE rule found in the pod's image, or it only considers it for health checks but not for user traffic)?

**Comment by mkarg** [comment]
Could anybody please help me? I am stuck and desperate! 😢

**Comment by mkarg** [comment]
Thank you, @liangyuanpeng! In fact, I made it work, but I used `hostPort` instead of `nodePort` (as told [in the kind docs](https://kind.sigs.k8s.io/docs/user/configuration/#extra-port-mappings)).

For those interested in a *complete and successfully working* code example, I am answering myself hereby. The following is the complete code needed to publish a pod's port 8080 (this is what the pod's image tells docker by `EXPOSE`) at the host's port 8888 (which is arbitrarily chosen). I am deliberately using different port numbers here to make perfectly clear which port number belongs to which object (pod, control-plane-container, host).
```yml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
  extraPortMappings:
  - containerPort: 8989
    hostPort: 8888
```
The mapping 8888:8989 tells docker to forward traffic from localhost:8888 to the control-plane-container's port 8989. `hostPort` and `containerPort` are exactly what they say: The *host*'s and the *control-plane-container*'s ports.
```yml
apiVersion: v1
kind: Pod
metadata:
  name: helloworld
spec:
  containers:
  - name: helloworld
    image: <image>
    ports:
    - containerPort: 8080
      hostPort: 8989
```
The mapping 8989:8080 tells kubernetes to forward traffic from the control-plane-container's port 8989 to the pod's port 8080 (to be more precide: the pod has no port itself, it is the port 8080 of the helloworld container running on that pod). `hostPort` and `containerPort` are rather confusingly used here (at least for k8s beginners) but make perfectly sense from a k8s view: `hostPort` here actually is the *control-plane-container's* port, while `containerPort` is the *pod*'s port. The first makes sense once we accept that we do kind (kubernetes-in-docker) here, which is a bit like dind (docker-in-docker): A pod could be seen as another docker instance running a *container* (the helloworld image which `EXPOSE`s port 8080), and the pod "sees" the control-plane-container as the node itself runs upon, hence is like a virtual *host*. So in the end, the names *do* make sense, once we got the big picture. 😅 

Now that we see the big picture and understood what-is-what, we can simply replace *all* port numbers by 8080 and it still will work like a charm. 😄
